# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import datetime
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

# check directory
os.getcwd()

'C:\\Users\\mdurr\\Dropbox\\Springboard\\Projects\\GuidedCapstone\\5Modeling'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
# Change directory
path = 'C:\\Users\\mdurr\\Dropbox\\Springboard\\Projects\\GuidedCapstone\\2DataWrangling'
os.chdir(path)

In [3]:
# Confirm
os.listdir()

['.ipynb_checkpoints',
 '1588204754_Step_2.zip',
 'data',
 'figures',
 'GuidedCapstone_Step2HL.ipynb',
 'models',
 'step2_output.csv',
 'step3_output.csv',
 'updated_ski_data.csv']

In [4]:
data = pd.read_csv('step3_output.csv')

# Check
data.head()

Unnamed: 0  Unnamed: 0.1                 Name    state  summit_elev  \
0           0             0       Alyeska Resort   Alaska         3939   
1           1             1  Eaglecrest Ski Area   Alaska         2600   
2           2             2     Hilltop Ski Area   Alaska         2090   
3           3             3     Arizona Snowbowl  Arizona        11500   
4           4             4  Sunrise Park Resort  Arizona        11100   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  ...  Snow Making_ac  \
0           2500      1        0.0          0          2  ...           113.0   
1           1540      0        0.0          0          0  ...            60.0   
2            294      0        0.0          0          0  ...            30.0   
3           2300      0        0.0          1          0  ...           104.0   
4           1800      0        0.0          0          1  ...            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  difference  clusters  
0              150.0           550.0         0.0         0  
1               90.0             0.0         0.0         0  
2              152.0            30.0         0.0         0  
3              122.0             0.0         0.0         1  
4              104.0            80.0         0.0         1  

[5 rows x 29 columns]

In [5]:
# Drop extra index columns
data.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis=1, inplace=True)
data.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  Snow Making_ac  daysOpenLastYear  \
0          0          2     2       0  ...           113.0             150.0   
1          0          0     0       0  ...            60.0              45.0   
2          0          0     0       1  ...            30.0             150.0   
3          1          0     2       2  ...           104.0             122.0   
4          0          1     2       3  ...            80.0             115.0   

   yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0       60.0            669.0          65.0          85.0              150.0   
1       44.0            350.0          47.0          53.0               90.0   
2       36.0             69.0          30.0          34.0              152.0   
3       81.0            260.0          89.0          89.0              122.0   
4       49.0            250.0          74.0          78.0              104.0   

   NightSkiing_ac  difference  clusters  
0           550.0         0.0         0  
1             0.0         0.0         0  
2            30.0         0.0         0  
3             0.0         0.0         1  
4            80.0         0.0         1  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
data = pd.concat([data.drop('state', axis=1), pd.get_dummies(data['state'])], axis=1)

In [7]:
# Check
data.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [8]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = data.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = data['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [9]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [10]:
# Investigate & Check
print(X_train.shape)
print(X_test.shape)
print(X.shape)

(247, 59)
(83, 59)
(330, 59)


Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

print("Explained Variance score: {:.2f}".format(explained_variance_score(y_test, y_pred)))
print("Mean Absolute error: {:.2f}".format(mean_absolute_error(y_test, y_pred)))

Explained Variance score: 0.94
Mean Absolute error: 4.85


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [14]:
print(lm.intercept_)

11.916043681962392


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [15]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
features = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(features['Coefficient'].sort_values(ascending=False)[0:10])

Tennessee         19.369347
North Carolina    12.731935
Maryland          10.234037
West Virginia      9.959850
Montana            9.446389
Massachusetts      7.996933
New Mexico         7.437658
Idaho              7.236024
Virginia           6.839755
South Dakota       6.744792
Name: Coefficient, dtype: float64


**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [16]:
# List of states to remove
states = ['Alaska', 'Arizona',
       'California', 'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming']
X_nostates = X.drop(states, axis=1)

# Scale X_nostates
scaler = StandardScaler().fit(X_nostates)
X_scaled=scaler.transform(X_nostates) 

# Split into train/test sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [17]:
# Create and fit model
lm2 = linear_model.LinearRegression()
model = lm2.fit(X_train,y_train)

# Test model
y_pred = lm2.predict(X_test)

In [18]:
# Check scoring of new model
print("Model score: {:.2f}".format(lm2.score(X_test, y_test)))
print("Explained Variance score: {:.2f}".format(explained_variance_score(y_test, y_pred)))
print("Mean Absolute error: {:.2f}".format(mean_absolute_error(y_test, y_pred)))

Model score: 0.93
Explained Variance score: 0.93
Mean Absolute error: 5.16


In [19]:
# Check top features of the new model without states
features = pd.DataFrame(abs(lm2.coef_), X_nostates.columns, columns=['Coefficient'])
print(features['Coefficient'].sort_values(ascending=False))


AdultWeekday         19.825672
clusters              5.957418
summit_elev           4.636055
Runs                  1.918079
triple                1.397295
surface               1.391804
quad                  1.081563
averageSnowfall       1.076867
vertical_drop         1.029008
fastQuads             0.990255
NightSkiing_ac        0.712154
Snow Making_ac        0.680674
total_chairs          0.608217
LongestRun_mi         0.516779
daysOpenLastYear      0.464370
SkiableTerrain_ac     0.431801
fastSixes             0.304101
yearsOpen             0.278414
double                0.230603
fastEight             0.216632
trams                 0.167398
TerrainParks          0.057521
projectedDaysOpen     0.054654
difference            0.000000
Name: Coefficient, dtype: float64


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [20]:
removals = ['summit_elev'] 
# base_elev was dropped in EDA step as was over 95% correlated with summit_elev
X_final = X_nostates.drop(removals, axis=1)

# Scale X_nostates
scaler = StandardScaler().fit(X_final)
X_scaled=scaler.transform(X_final) 

# Split into train/test sets
X_train, X_test, y_train, y_test = train_test_split(X_final, y, test_size=0.25, random_state=1)

In [21]:
# Create and fit model
lm3 = linear_model.LinearRegression()
model = lm3.fit(X_train,y_train)

# Test model
y_pred = lm3.predict(X_test)

In [22]:
# Check scoring of new model

print("Explained Variance score: {:.2f}".format(explained_variance_score(y_test, y_pred)))
print("Mean Absolute error: {:.2f}".format(mean_absolute_error(y_test, y_pred)))

Explained Variance score: 0.93
Mean Absolute error: 5.16


In [23]:
# Check top features of the new model without states
features = pd.DataFrame(abs(lm3.coef_), X_final.columns, columns=['Coefficient'])
print(features['Coefficient'].sort_values(ascending=False))

clusters             5.477322
trams                0.911139
surface              0.880499
AdultWeekday         0.839895
fastEight            0.786663
quad                 0.604507
triple               0.567568
total_chairs         0.404595
LongestRun_mi        0.359618
fastSixes            0.346639
TerrainParks         0.179801
double               0.147026
fastQuads            0.037882
Runs                 0.030873
daysOpenLastYear     0.028533
projectedDaysOpen    0.013295
averageSnowfall      0.006386
NightSkiing_ac       0.004017
Snow Making_ac       0.002801
yearsOpen            0.002181
vertical_drop        0.002161
SkiableTerrain_ac    0.000089
difference           0.000000
Name: Coefficient, dtype: float64


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.94 | 4.85 |-|
| Model 2. | 0.93 | 6.16 |'state'|
| Model 3. | 0.93 | 5.16|'state','summit_elev','base_elev'|

Model Selection:

I will use the first model. Although it contains features outside of management's control - such as the state - the park's location will be helpful to factor into pricing decisions. 

For example: although the park can't be moved, if other parks in the same state are chargining significantly different prices for relatively similiar experiences, that is information that will be helpful in making our own pricing decisions.